### PR TITLE
 Add mediaCannotPlay and onTrackPlaybackFailure

### PR DESCRIPTION
**Pull request by danielr18**
Solves #352.

I documented `mediaCannotPlay`, but not onTrackPlaybackFailure, since the markdown file for playerContextProvider is empty and the Readme options look outdated.

### CODE REVIEW

**Comment by danielr18** [comment]
Don't know if we should pass the activeTrackIndex or something to onTrackPlaybackFailure.

File: packages/core/src/PlayerContextProvider.js [hunk 1]
```diff
@@ -499,6 +491,35 @@ export class PlayerContextProvider extends Component {
...
+    // cancel playback and re-scan new sources
+    this.media.load();
+  }
+
+  onTrackPlaybackFailure(event) {
+    this.setState({
+      mediaCannotPlay: true
+    });
+    if (this.props.onTrackPlaybackFailure) {
+      this.props.onTrackPlaybackFailure(event);
```
Comment (benwiley4000) [inline packages/core/src/PlayerContextProvider.js]: I think this should be:
```js
this.props.onTrackPlaybackFailure(
  this.props.playlist[this.state.activeTrackIndex],
  this.state.activeTrackIndex,
  event
);
```

File: packages/core/src/PlayerContextProvider.js [hunk 2]
```diff
@@ -345,7 +336,8 @@ export class PlayerContextProvider extends Component {
     // if not, then load the first track in the new playlist, and pause.
     return {
       ...baseNewState,
-      ...getGoToTrackState({ prevState, index: 0, shouldPlay: false })
+      ...getGoToTrackState({ prevState, index: 0, shouldPlay: false }),
+      mediaCannotPlay: false
```
Comment (benwiley4000) [inline packages/core/src/PlayerContextProvider.js]: This looks good. Can we also set `mediaCannotPlay: false` above when `!isPlaylistValid(newPlaylist)`?

Comment (benwiley4000) [reply packages/core/src/PlayerContextProvider.js]: Just kidding ignore this - I realized that's part of the `defaultState`.

File: packages/core/src/PlayerContextProvider.js [hunk 3]
```diff
@@ -166,6 +153,10 @@ export class PlayerContextProvider extends Component {
     this.videoHostOccupiedCallbacks = new Map();
     this.videoHostVacatedCallbacks = new Map();
 
+    // bind internal methods
+    this.setMediaElementSources = this.setMediaElementSources.bind(this);
```
[the remark this answers is not in the document]
Comment (danielr18) [reply packages/core/src/PlayerContextProvider.js]: `setMediaElementSources` method uses `this`, wouldn't that require it to be bound?

[the remark this answers is not in the document]
Comment (benwiley4000) [reply packages/core/src/PlayerContextProvider.js]: It's not passed anywhere as a callback so it's bound at call time when we call `this.setMediaElementSources(...)`.

File: packages/core/src/PlayerContextProvider.js [hunk 4]
```diff
@@ -499,6 +491,35 @@ export class PlayerContextProvider extends Component {
...
+      this.media.removeChild(firstChild);
+    }
+    // add new sources
+    for (const source of sources) {
+      const sourceElement = document.createElement('source');
+      sourceElement.src = source.src;
+      if (source.type) {
+        sourceElement.type = source.type;
+      }
+      sourceElement.addEventListener('error', this.onTrackPlaybackFailure);
```
[the remark this answers is not in the document]
Comment (danielr18) [reply packages/core/src/PlayerContextProvider.js]: Wouldn't it be garbage collected as you said in the issue?

https://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory

Except in old browsers.

[the remark this answers is not in the document]
Comment (danielr18) [reply packages/core/src/PlayerContextProvider.js]: However, doesn't hurt to remove them, check out the last commit.

File: packages/core/src/PlayerContextProvider.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielr18) [reply packages/core/src/PlayerContextProvider.js]: Ahh, yes! Just removed it.

File: packages/core/src/PlayerContextProvider.js (hunk 1 above)
Comment (danielr18) [reply packages/core/src/PlayerContextProvider.js]: Good, done.

File: packages/core/src/PlayerContextProvider.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (benwiley4000) [reply packages/core/src/PlayerContextProvider.js]: Normally yes, but if someone retained an outside ref to the media element then I want to make sure we don't have callbacks belonging to an unmounted component getting called.

**Comment by benwiley4000** [comment]
Released in v2.0.0-alpha.23.

**Pull request merged**